### ISSUE TITLE
PrepareHeaderForMatch Complier Issue after Upgrade to v25

### ISSUE THREAD

**Issue by ExpenseOnce** [JoshClose/CsvHelper#1727]
**Description**
I just upgraded to the latest CsvHelper v25, the project is built using C# language v7.3 with .NET v4.7.2. Building and testing locally works ok but when I run the pipeline and build the solution on Azure, I'm getting a compiler issue.

`Property, indexer, or event 'PrepareHeaderForMatchArgs.Header' is not supported by the language; try directly calling accessor methods 'PrepareHeaderForMatchArgs.get_Header()' or 'PrepareHeaderForMatchArgs.set_Header(?)'`

It's perhaps the way I'm using PrepareHeaderForMatch in the following snippet:

```
  using (StreamReader r = new StreamReader(file.InputStream))
            using (var csv = new CsvReader(r, CultureInfo.InvariantCulture))
            {
                csv.Context.TypeConverterCache.AddConverter<Decimal>(new DecimalConverter());
                csv.Context.TypeConverterCache.AddConverter<DateTime>(new DateConverter());
                csv.Context.RegisterClassMap(mapper);

                **csv.Context.Configuration.PrepareHeaderForMatch = args => args.Header.ToLowerInvariant();**

                data = csv.GetRecords<BankDetailFields>().ToList();
            }

```
**Describe alternatives**
I've checked and made sure its using C# v7.3 on deployment. I also tried a few variants of the syntax but I'm a little stuck.

**Additional context**
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs1546

Has anyone come across this issue, or there something wrong with my approach.

Many thanks.

**Comment by JoshClose** [comment]
The configuration is read only. You need to set that when creating the configuration and pass it into the `CsvReader` constructor.

```cs
var config = new CsvConfiguration(CultureInfo.InvariantCulture)
{
	PrepareHeaderForMatch = args => args.Header.ToLowerInvariant()
};
using (var csv = new CsvReader(r, config))
```

You shouldn't be able to set it locally.

https://github.com/JoshClose/CsvHelper/blob/86d21c44d576d8bd37f910421b6d8e990b8f2b1d/src/CsvHelper/Configuration/IReaderConfiguration.cs#L59

**Comment by ExpenseOnce** [comment]
Ah right, thank you Josh. I've corrected the implementation but still getting the CS1546 exception if I include `PrepareHeaderForMatch = args => args.Header.ToLowerInvariant() `. 

I'll try it in a new project and narrow down the problem.

**Comment by JeffKulon** [comment]
Having similar issues, regarding varying build environments.  PrepareHeaderForMatch**Args** ( the type for `args` in the snippets above ) uses C#9 init-only properties, and that could be the cause of the CS1546.  In the original post, the reported error is identifying `PrepareHeaderForMatchArgs.Header` specifically, not CsvConfiguration.

**Comment by JoshClose** [comment]
Using `init` should be fine. You do need Visual Studio 2019 16.8 or MSBuild 2019 16.8 or later for it to work though I believe. If not, it will show up as `get_Header` I believe. I thought it would be ok since it wasn't requiring the user to use `init`, but apparently not.

Sigh... maybe I'll just have to make them settable or add constructors again. :(

**Comment by JoshClose** [comment]
This is fixed in 26.0.0 on NuGet. https://www.nuget.org/packages/CsvHelper/26.0.0

I removed all `init` properties.
